libcudacxx/include/cuda/std/detail/libcxx/include/__config

```diff
@@ -1439,11 +1439,20 @@ typedef __char32_t char32_t;
 #endif
 #endif
 
+// We want to hide defaulted SMF but those cannot be marked as __host__ __device__
+#ifndef _LIBCUDACXX_HIDE_FROM_ABI_DEFAULTED
+#  if _LIBCUDACXX_HIDE_FROM_ABI_PER_TU
+#    define _LIBCUDACXX_HIDE_FROM_ABI_DEFAULTED inline _CCCL_ATTRIBUTE_HIDDEN _LIBCUDACXX_INTERNAL_LINKAGE
+#  else
+#    define _LIBCUDACXX_HIDE_FROM_ABI_DEFAULTED inline _CCCL_ATTRIBUTE_HIDDEN _LIBCUDACXX_EXCLUDE_FROM_EXPLICIT_INSTANTIATION
+#  endif
+#endif
+
 #ifndef _LIBCUDACXX_HIDE_FROM_ABI
 #  if _LIBCUDACXX_HIDE_FROM_ABI_PER_TU
-#    define _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_HIDDEN _LIBCUDACXX_INTERNAL_LINKAGE
+#    define _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_HIDE_FROM_ABI_DEFAULTED _LIBCUDACXX_HOST_DEVICE
```

Comment:
question:

We used to have:

```cpp
#ifdef __cuda_std__
#  define _LIBCUDACXX_INLINE_VISIBILITY \
        __host__ __device__
#else
#  if _LIBCUDACXX_HIDE_FROM_ABI_PER_TU
#    define _LIBCUDACXX_INLINE_VISIBILITY \
        __attribute__ ((__visibility__("hidden")))
#  else 
#    define _LIBCUDACXX_INLINE_VISIBILITY \
        __attribute__ ((__exclude_from_explicit_instantiation__))
#  endif
#endif // __cuda_std__

inline _LIBCUDACXX_INLINE_VISIBILITY  void foo();
_LIBCUDACXX_INLINE_VISIBILITY  void bar();
```

This PR suggests:

```cpp
#if _LIBCUDACXX_HIDE_FROM_ABI_PER_TU
#  define _LIBCUDACXX_HIDE_FROM_ABI 
      __host__ __device__                        \
      inline                                     \
      __attribute__ ((__visibility__("hidden"))) \
      __attribute__ ((internal_linkage))
#else
#  define _LIBCUDACXX_HIDE_FROM_ABI 
      __host__ __device__                        \
      inline                                     \
      __attribute__ ((__visibility__("hidden"))) \
      __attribute__ ((__exclude_from_explicit_instantiation__))
#endif

_LIBCUDACXX_HIDE_FROM_ABI void foo();
_LIBCUDACXX_HIDE_FROM_ABI void bar();
```

In libc++:

```cpp
#define _LIBCPP_HIDE_FROM_ABI \
  __attribute__ ((__visibility__("hidden"))) \
  __attribute__((__exclude_from_explicit_instantiation__)) 

inline _LIBCPP_HIDE_FROM_ABI void foo();
_LIBCPP_HIDE_FROM_ABI void bar();
```

The change seems quite different from both what we used to have and what libcu++ has. Could you share your understanding of why this change is safe?

Reply:
So there is no question that this is a functional change, but one that is intended.

The difference is that outside of clang we do not use `__exclude_from_explicit_instantiation__` or `internal_linkage` but we fall back to `__attribute__((allways_inline))` which only works if the function is marked `inline`. 

Now the sad issue is that nvcc also complains about duplicated `inline` attributes which forces us to put `inline` into the macro